ntoskrnl/io/pnpmgr/pnpenum.c

```diff
@@ -0,0 +1,339 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Device enumeration functions
+ * COPYRIGHT:   2018 Vadim Galyant
+ */
+
+#include <ntoskrnl.h>
+
+#define NDEBUG
+#include <debug.h>
+
+extern KSPIN_LOCK IopPnPSpinLock;
+extern LIST_ENTRY IopPnpEnumerationRequestList;
+extern KEVENT PiEnumerationFinished;
+
+// Request types for PIP_ENUM_REQUEST
+typedef enum _PIP_ENUM_TYPE
+{
+    PipEnumAddBootDevices,
+    PipEnumAssignResources,
+    PipEnumGetSetDeviceStatus,
+    PipEnumClearProblem,
+    PipEnumInvalidateRelationsInList,
+    PipEnumHaltDevice,
+    PipEnumBootDevices,
+    PipEnumDeviceOnly,
+    PipEnumDeviceTree,
+    PipEnumRootDevices,
+    PipEnumInvalidateDeviceState,
+    PipEnumResetDevice,
+    PipEnumIoResourceChanged,
+    PipEnumSystemHiveLimitChange,
+    PipEnumSetProblem,
+    PipEnumShutdownPnpDevices,
+    PipEnumStartDevice,
+    PipEnumStartSystemDevices
+} PIP_ENUM_TYPE;
+
+typedef struct _PIP_ENUM_REQUEST
+{
+    LIST_ENTRY RequestLink;
+    PDEVICE_OBJECT DeviceObject;
+    PIP_ENUM_TYPE RequestType;
+    UCHAR ReorderingBarrier;
+    ULONG_PTR RequestArgument;
+    PKEVENT CompletionEvent;
+    NTSTATUS *CompletionStatus;
+} PIP_ENUM_REQUEST, *PPIP_ENUM_REQUEST;
+
+BOOLEAN PipEnumerationInProgress;
+WORK_QUEUE_ITEM PipDeviceEnumerationWorkItem;
+
+
+VOID
+NTAPI
+PipEnumerationWorker(_In_ PVOID Context)
+{
+    PDEVICE_OBJECT DeviceObject;
+    PDEVICE_NODE DeviceNode;
+    PPIP_ENUM_REQUEST Request;
+    BOOLEAN IsDereferenceObject;
+    BOOLEAN IsBootProcess = FALSE;
+    BOOLEAN IsAssignResources = FALSE;
+    KIRQL OldIrql;
+    NTSTATUS Status;
+
+    PpDevNodeLockTree(1);
+
+    while (TRUE)
+    {
+        Status = STATUS_SUCCESS;
+        IsDereferenceObject = TRUE;
+
+        KeAcquireSpinLock(&IopPnPSpinLock, &OldIrql);
+
+        Request = CONTAINING_RECORD(IopPnpEnumerationRequestList.Flink,
+                                    PIP_ENUM_REQUEST,
+                                    RequestLink);
+
+        if (IsListEmpty(&IopPnpEnumerationRequestList))
+        {
+            break;
+        }
+
+        RemoveHeadList(&IopPnpEnumerationRequestList);
+        KeReleaseSpinLock(&IopPnPSpinLock, OldIrql);
+
+        if (Request)
+        {
+            InitializeListHead(&Request->RequestLink);
+
+            //FIXME: Check ShuttingDown
+
+            DeviceObject = Request->DeviceObject;
+            ASSERT(DeviceObject);
+
+            DeviceNode = IopGetDeviceNode(DeviceObject);
+            ASSERT(DeviceNode);
+
+            if (DeviceNode->State == DeviceNodeDeleted)
+            {
+                DPRINT("PipEnumerationWorker: DeviceNode->State == DeviceNodeDeleted\n");
+                Status = STATUS_UNSUCCESSFUL;
+                goto NextRequest;
```

Comment:
I would suggest to modify this `if (DeviceNode->State ...)` check in order to avoid the goto NextRequest.